= About the GWAS Catalog

== About the GWAS Catalog

The GWAS Catalog is provided jointly by the http://www.genome.gov[National Human Genome Research Institute (NHGRI)] and the http://www.ebi.ac.uk[European Bioinformatics Institute (EMBL-EBI)].
It is a quality controlled, manually curated, literature-derived collection of all published http://en.wikipedia.org/wiki/Genome-wide_association_study[genome-wide association studies] assaying at least 100,000 http://en.wikipedia.org/wiki/Single-nucleotide_polymorphism[SNPs] and all SNP-trait associations with p-values < 1.0 E-5 _(Hindorff et al., 2009)_.

The Catalog also publishes the iconic GWAS diagram of all SNP-trait associations, with p-values < 5.0 x 10^-8^, mapped onto the human genome by chromosomal locations and displayed on the human karyotype.
The diagram is released quarterly and the latest version of the diagram is made available on our website in PDF format and as a PowerPoint slide.

== Curation process

The GWAS Catalog data is extracted from the literature. Extracted information includes publication information, study cohort information such as cohort size, country of recruitment and subject ethnicity, and SNP-disease association information including SNP identifier (i.e. RSID), p-value, gene and risk allele.
Each study is also assigned a trait that best represents the phenotype under investigation.
When multiple traits are analysed in the same study either multiple entries are created, or individual SNPs are annotated with their specific traits.
Traits are used both to query and visualise the data in the Catalog's web form and diagram-based query interfaces.

Data extraction and curation for the GWAS Catalog is an expert activity; each step is performed by scientists supported by a web-based tracking and data entry system which allows multiple curators to search, annotate, verify and publish the Catalog data.
Papers that qualify for inclusion in the Catalog are identified through weekly PubMed searches.
They then undergo two levels of curation.
First all data, including association information for SNPs, traits and general information about the study, are extracted by one curator.
A second curator then performs an additional round of curation to double-check the accuracy and consistency of all the information.
Finally, an automated pipeline run at the NCBI performs checks for missing information, such as SNP identifiers, resolution of the citation in PubMed, existence of SNPs in dbSNP, confirmation of gene names vs. NCBI's Gene database, confirmation that the reported SNP and gene are in the same chromosomal region, as well as adding additional information such as the SNP's base pair and cytogenetic location in the current human genome reference assembly.
This information is then used for queries and in the production of the karyotype diagram.

== Citing the GWAS Catalog

===  How to cite the NHGRI GWAS Catalog:

_Hindorff LA, MacArthur J (European Bioinformatics Institute), Morales J (European Bioinformatics Institute), Junkins HA, Hall PN, Klemm AK, and Manolio TA._
The NHGRI-EBI Catalog of published genome-wide association studies.
Available at: www.genome.gov/gwastudies. Accessed [date of access].

=== How to cite the NHGRI GWAS Catalog paper:

_Welter D, MacArthur J, Morales J, Burdett T, Hall P, Junkins H, Klemm A, Flicek P, Manolio T, Hindorff L, and Parkinson H._
The NHGRI GWAS Catalog, a curated resource of SNP-trait associations.
Nucleic Acids Research, 2014, Vol. 42 (Database issue): D1001-D1006.

== Related resources

==== GWAS Webinar
On Thursday, July 18th, 2013, the Division of Genomic Medicine held http://www.genome.gov/27554296[a webinar] to highlight current uses and explore  priorities and future directions for the GWAS catalog.

==== Various resources & projects
http://www.ensembl.org[Ensembl Genome Browser]

http://www.gen2phen.org/[GEN2PHEN Project]

http://geneticassociationdb.nih.gov/[Genetic Association Database]

http://www.geneticsofgeneexpression.org/network/[Human Gene Coexpression Database]

http://www.hapmap.org/[International HapMap Project]

http://www.phenxtoolkit.org[PhenX Toolkit (consensus measures of Phenotypes and eXposures)]

http://genome.ucsc.edu/cgi-bin/hgGateway[Genome Browser Gateway]

==== HuGE Navigator (Centers for Disease Control and Prevention)
http://hugenavigator.net/HuGENavigator/gWAHitStartPage.do[GWAS Integrator]

http://www.hugenavigator.net/HuGENavigator/startPagePubLit.do[HuGE Literature Finder]

http://www.hugenavigator.net/HuGENavigator/startPageMapper.do[Variant Name Mapper]


==== National Center for Biotechnology Information (NCBI) Links
http://www.ncbi.nlm.nih.gov/projects/SNP/[dbSNP]

http://www.ncbi.nlm.nih.gov/sites/entrez?db=gene[Entrez Gene]

http://www.ncbi.nlm.nih.gov/Genbank/[GenBank]

http://www.ncbi.nlm.nih.gov/sites/entrez/[PubMed]

http://www.ncbi.nlm.nih.gov/gap/PheGenIUCSC[Phenotype-Genotype Integrator] 



== Credits &amp; acknowledgements

Font generated by http://www.flaticon.com[flaticon.com] under http://creativecommons.org/licenses/by/3.0/[CC BY license].
The authors are http://picol.org[Picol].
