= Description of column headings for mapping file

== Description of column headings for mapping file

DATE ADDED TO CATALOG: Date added to catalog

PUBMEDID: PubMed identification number

FIRST AUTHOR: Last name and initials of first author

DATE: Publication date (online (epub) date if available)

JOURNAL: Abbreviated journal name

LINK: PubMed URL

STUDY: Title of paper

DISEASE/TRAIT: Disease or trait examined in study

INITIAL SAMPLE DESCRIPTION: Sample size and ancestry description for stage 1 of GWAS (summing across multiple Stage 1 populations, if applicable)

REPLICATION SAMPLE DESCRIPTION: Sample size and ancestry description for subsequent replication(s) (summing across multiple populations, if applicable)

REGION: Cytogenetic region(s) associated with rs number. A comma separated list is included if the rs number maps to multiple locations.

CHR_ID: Chromosome number(s) associated with rs number. A comma separated list is included if the rs number maps to multiple locations.

CHR_POS: Chromosomal position(s) associated with rs number. A comma separated list is included if the rs number maps to multiple locations.

REPORTED GENE(S): Gene(s) reported by author

ENTREZ_MAPPED_GENE: Gene(s) mapped to the strongest SNP (NCBI data retrieved from Ensembl). If the SNP is located within a gene that gene is listed. A comma separated list of genes is included if the strongest SNP maps to multiple genes. If the SNP is intergenic, the closest upstream and downstream genes are listed, separated by a hyphen.

ENSEMBL_MAPPED_GENE: Gene(s) mapped to the strongest SNP (Ensembl data retrieved from Ensembl). If the SNP is located within a gene that gene is listed. A comma separated list of genes is included if the strongest SNP maps to multiple genes. If the SNP is intergenic, the closest upstream and downstream genes are listed, separated by a hyphen.

ENTREZ_UPSTREAM_GENE_ID: Entrez Gene ID for nearest upstream gene to rs number, if not within gene (NCBI data retrieved from Ensembl)

ENTREZ_DOWNSTREAM_GENE_ID: Entrez Gene ID for nearest downstream gene to rs number, if not within gene (NCBI data retrieved from Ensembl)

ENSEMBL_UPSTREAM_GENE_ID: Ensembl Gene ID for nearest upstream gene to rs number, if not within gene (Ensembl data retrieved from Ensembl)

ENSEMBL_DOWNSTREAM_GENE_ID: Ensembl Gene ID for nearest downstream gene to rs number, if not within gene (Ensembl data retrieved from Ensembl)

SNP_GENE_IDS_ENTREZ: Entrez Gene ID, if rs number within gene; multiple genes denotes overlapping transcripts (NCBI data retrieved from Ensembl)

SNP_GENE_IDS_ENSEMBL: Ensembl Gene ID, if rs number within gene; multiple genes denotes overlapping transcripts (Ensembl data retrieved from Ensembl)

ENTREZ_UPSTREAM_GENE_DISTANCE: distance in kb for nearest upstream gene to rs number, if not within gene (NCBI data retrieved from Ensembl)

ENTREZ_DOWNSTREAM_GENE_DISTANCE: distance in kb for nearest downstream gene to rs number, if not within gene (NCBI data retrieved from Ensembl)

ENSEMBL_UPSTREAM_GENE_DISTANCE:	distance in kb for nearest upstream gene to rs number, if not within gene (Ensembl data retrieved from Ensembl)

ENSEMBL_DOWNSTREAM_GENE_DISTANCE:  distance in kb for nearest downstream gene to rs number, if not within gene (Ensembl data retrieved from Ensembl)

STRONGEST SNP-RISK ALLELE: SNP(s) most strongly associated with trait + risk allele (? for unknown risk allele). May also refer to a haplotype.

SNPS: Strongest SNP; if a haplotype it may include more than one rs number (multiple SNPs comprising the haplotype)

MERGED: denotes whether the SNP has been merged into a subsequent rs record (0 = no; 1 = yes)

SNP_ID_CURRENT: current rs number (will differ from strongest SNP when merged = 1)

CONTEXT: SNP functional class

INTERGENIC_ENTREZ: denotes whether SNP is in intergenic region (0 = no; 1 = yes; NCBI data retrieved from Ensembl)

INTERGENIC_ENSEMBL: denotes whether SNP is in intergenic region (0 = no; 1 = yes; Ensembl data retrieved from Ensembl)

RISK ALLELE FREQUENCY: Reported risk/effect allele frequency associated with strongest SNP in controls (if not available among all controls, among the control group with the largest sample size). If the associated locus is a haplotype the haplotype frequency will be extracted.

P-VALUE: Reported p-value for strongest SNP risk allele (linked to dbGaP Association Browser). Note that p-values are rounded to 1 significant digit (for example, a published p-value of 4.8 x 10-7 is rounded to 5 x 10-7).

PVALUE_MLOG: -log(p-value)

P-VALUE (TEXT): Information describing context of p-value (e.g. females, smokers).

OR or BETA: Reported odds ratio or beta-coefficient associated with strongest SNP risk allele. Note that if an OR <1 is reported this is inverted, along with the reported allele, so that all ORs included in the Catalog are >1. Appropriate unit and increase/decrease are included for beta coefficients.

95% CI (TEXT): Reported 95% confidence interval associated with strongest SNP risk allele, along with unit in the case of beta-coefficients. If 95% CIs are not published, we estimate these using the standard error, where available.

PLATFORM (SNPS PASSING QC): Genotyping platform manufacturer used in Stage 1; also includes notation of pooled DNA study design or imputation of SNPs, where applicable

_Last updated: 12 January 2016_