= Pilots
:imagesdir: ./images
:data-uri:

== Pilots to investigate extending Catalog scope and alternative methods of data acquisition

We have carried out several pilots to investigate extending Catalog scope and alternative methods of data acquisition with a view to incorporating these into future Catalog improvements. Each of these pilots has been designed to allow us to estimate the feasibility of the approach, resources needed to implement and support this, and obtain feedback from users.

=== Author deposition pilot

To support scaling of curation alternative methods of data acquisition have been explored. The pilot involved:

* Designing a prototype deposition system was using online deposition forms (created in Cognito Forms, see figures below) with templates for sample and association results (created in Google Sheets).
* Emailing 115 authors of 79 selected GWAS publications (53 whole-genome array and 26 targeted array) and inviting them to submit data using the test submission system. The deposition pilot was also advertised on Twitter, which was re-tweeted to over 20,000 followers.
* We received an uptake of 10% from direct emails, with no uptake from Twitter.
* Feedback from submitters, along with review of the submitted data, indicated that the format of deposition is easy to understand and allows authors to submit all relevant data with a high level of accuracy and rapidly.
* Even a small rate of deposition represents a gain for the Catalog as it removes a lengthy paper reading and literature extraction step and the quality of deposited data is high. It should also be noted that retrospective deposition is not our preferred model and reduces take up.

{empty} +

_Author deposition prototype, main page_

image::author_sub_pilot_main.png[Author submission,500,500,align="center"]

{empty} +

_Author deposition prototype, sample descriptions page_

image::author_sub_samples.jpg[Sample descriptions,500,500,align="center"]

{empty} +

_Author deposition prototype, associations upload page_

image::author_sub_associations.png[Associations,500,500,align="center"]

{empty} +


Data submitted as part of the author deposition pilot can be found link:ftp://ftp.ebi.ac.uk/pub/databases/gwas/pilot_data/author_submission_201608/[on our FTP server].



=== Targeted array pilot

Feedback from our users has indicated a high demand for targeted array studies to be included in the GWAS Catalog. 

In 2016 we carried out a pilot on the inclusion of targeted array based GWAS in the Catalog involving:

* Designing and carrying out a literature search to identify all studies that carry out genotyping and association analysis of at least 100,000 SNPs.
* This identified approximately 150 targeted array studies that have been published up to June 2016 (including 62 Immunochip, 40 Metabochip and 36 exome arrays)
* Curation of a representative set of these studies indicated that inclusion would require minimal changes to Catalog infrastructure and has enabled us to estimate the additional curation resources required to include these in the future.

Data from the studies curated as part of the targeted array can be found link:ftp://ftp.ebi.ac.uk/pub/databases/gwas/pilot_data/targeted_array_201607/[on our FTP server].

Following on from this, we are currently piloting the inclusion of targeted array studies in the Catalog. Prioritisation of targeted and exome array studies for inclusion in the Catalog is now performed by 1) relevance of the trait analysed 2) user request, with Open Targets being the main user in this phase of the pilot.
In September 2017, Open Targets (www.opentargets.org) requested curation of fifty-five currently out-of-scope GWAS publications for inclusion in the GWAS Catalog.  Moreover, the GWAS Catalog team have preliminarily identified over 150 publications based on targeted or exome array analysis from 2012 to 2017. These will also be curated as part of the inclusion pilot.



=== Sequencing-based association studies

To ensure we keep up with the most current research, technologies and meets the needs of the user community we are investigating expanding the scope of the Catalog to include sequencing-based association studies. Currently the Catalog only includes studies which use array-based genotyping technologies.
Ten studies which include sequencing-based association analyses were curated as part of a pilot on expanding the scope of the NHGRI-EBI GWAS Catalog. Study level and association data are provided in the standard GWAS Catalog download format with additional columns for sequencing specific data.

Data is available on our FTP server link:ftp://ftp.ebi.ac.uk/pub/databases/gwas/pilot_data/sequencing_pilot_201710/[on our FTP server]

Study level and association data are provided in the standard GWAS Catalog download format. (gwas_catalog_v1.0.1-studies_r2017-10-09_Sequencing_pilot, gwas_catalog_v1.0.1-associations_r2017-10-09_Sequencing_pilot).

Additional study level data fields, specific to sequencing associations are provided as additional columns in gwas_catalog_v1.0.1-studies_r2017-10-09_Sequencing_pilot.  
These additional fields are: Genotyping technology (discovery/initial), Sequencing platform, Array platform, Sequencing coverage, Number of variants analysed, Genome build, Analysis level, For multiple variants - number of variants per test, Statistical tests used to generate reported p-values, Statistical test used for extracted p-values, Why this test was chosen for extraction, Types of variant included in analysis, P-value threshold stated by author.
Single variant/multi-variant analysis is indicated after the author name.  The absence of either indicates that entry is currently eligible and published in the standard GWAS Catalog.

Additional association level data is provided in the P-VALUE (TEXT) field of gwas_catalog_v1.0.1-associations_r2017-10-09_Sequencing_pilot, where 'TEST'=statistical test and 'NUM_VAR'=number of variants analysed.
