= Training & Related resources

=== GWAS Catalog training materials

==== EBI Train online

Check out link:http://www.ebi.ac.uk/training/online/course/gwas-catalog-exploring-snp-trait-associations[GWAS Catalog - Exploring SNP-trait associations] for an easily accessible interactive tutorial on the full range of features the GWAS Catalog has to offer.

==== Workshop materials

Training materials from workshops held by the GWAS Catalog team are available on link:ftp://ftp.ebi.ac.uk/pub/databases/gwas/training_materials/[our FTP server]. If you would like to make use of these slides, please cite the GWAS Catalog as detailed on the link:about[About] page.

==== GWAS Webinars

A 2017 link:http://www.ebi.ac.uk/training/online/course/nhgri-ebi-gwas-catalog-curated-resource-snp-trait-associations[webinar] on GWAS design and use of the GWAS Catalog, hosted by the GWAS Catalog Team. It covers topics such as Catalog content, data availability and access, search capability and interoperability with other resources.

A 2013 http://www.genome.gov/27554296[webinar] hosted by the NHGRI Division of Genomic Medicine, to highlight the utility and impact of the GWAS Catalog.


=== Publications

==== GWAS Catalog publications

Morales et al, 2017 link:https://www.biorxiv.org/content/early/2017/10/02/129395[A standardized framework for representation of ancestry data in genomics studies, with application to the NHGRI-EBI GWAS Catalog] (pre-print).

MacArthur et al, 2017. link:https://academic.oup.com/nar/article/45/D1/D896/2605751[The new NHGRI-EBI Catalog of published genome-wide association studies (GWAS Catalog)]. Nucleic Acids Research, Vol. 45 (Database issue): D896-D901.

Welter D et al, 2014 link:https://academic.oup.com/nar/article/42/D1/D1001/1062755[The NHGRI GWAS Catalog, a curated resource of SNP-trait associations]. Nucleic Acids Research, Vol. 42 (Database issue): D1001-D1006.

Hindorff LA, et al 2009 link:http://www.genome.gov/Pages/About/OD/NewsAndFeatures/PNASGWASOnlineCatalog.pdf[Potential etiologic and functional implications of genome-wide association loci for human diseases and traits]. Proc. Natl. Acad. Sci. USA, 2009, 106, 9362–9367.

==== Other background reading

link:https://www.yourgenome.org/stories/genome-wide-association-studies[A beginners guide to GWAS]

link:http://ontogenesis.knowledgeblog.org/66[What is an Ontology?]

Visscher et al, 2017 link:http://www.sciencedirect.com/science/article/pii/S0002929717302409?via%3Dihub[10 Years of GWAS Discovery: Biology, Function and Translation] AJHG Volume 101, Issue 1, 6 July 2017, Pages 5-22

Manolio, 2010 link:http://www.nejm.org/doi/full/10.1056/NEJMra0905980[Genomewide Association Studies and assessment of the risk of disease] NEJM 363:166-176



=== Resources used within the GWAS Catalog

For more information on how we use these resources, see our link:https://www.ebi.ac.uk/gwas/docs/methods[Methods page].

// tag::b-col-indv[]
[cols="1,2"]
|===

|link:https://www.ebi.ac.uk/efo/[Experimental Factor Ontology] (EFO)
|The ontology used for trait mapping in the GWAS Catalog

|link:http://www.ensembl.org/[Ensembl Genome Browser]
|Genome browser and data for vertebrate genomes, used in our variant mapping pipeline. Ensembl annotates genes, computes multiple alignments, predicts regulatory function and collects disease data

|link:https://www.ncbi.nlm.nih.gov/pubmed[PubMed]
|Literature search and citation annotation

|link:http://europepmc.org/[EuropePMC]
|Literature search and citation annotation

|link:https://www.cia.gov/library/publications/the-world-factbook/geos/lg.html[CIA Factbook]
|A rich source of population demographic data, used by curators

|link:https://www.ncbi.nlm.nih.gov/[NCBI]
|Data used for variant mapping in older versions of the Catalog

|===
// end::b-col-indv[]


=== Genome and phenome resources

// tag::b-col-indv[]
[cols="1,2"]
|===

|link:http://www.internationalgenome.org/[1000 Genomes]
|The largest public reference catalogue of worldwide human variation and genotype data

|http://www.hapmap.org/[International HapMap Project]
|The first public catalogue of human variation data, now superseded by 1000Genomes

|http://genome.ucsc.edu/cgi-bin/hgGateway[UCSC Genome Browser]
|Genome browser to interactively visualize genomic data

|link:http://www.phenxtoolkit.org/[PhenX toolkit]
|Consensus measures of Phenotypes and eXposures

|http://www.ncbi.nlm.nih.gov/projects/SNP/[dbSNP]
|Database of human single nucleotide polymorphisms

|https://www.ncbi.nlm.nih.gov/[National Centre for Biotechnology Information]
|Provides wide range of resources including Entrez gene, GenBank, PubMed and ClinVar

|===
// end::b-col-indv[]


=== Useful tools & data services


// tag::b-col-indv[]
[cols="1,2"]
|===

|https://www.ensembl.org/info/docs/tools/vep/index.html[Variant Effect Predictor] (VEP)
|Determines the effect of variants (SNPs, insertions, deletions, CNVs or structural variants) on genes, transcripts, and protein sequence, as well as regulatory regions

|http://www.phenoscanner.medschl.cam.ac.uk/phenoscanner[Phenoscanner]
|Tool to facilitate “phenome scans”, the cross-referencing of genetic variants with a broad range of phenotypes, using a range of data sources including GWAS, eQTL and metabolite data

|https://phgkb.cdc.gov/PHGKB/hNHome.action[HuGE Navigator]
|Knowledge base in human genome epidemiology, including information on population prevalence of genetic variants, gene-disease associations, gene-gene and gene- environment interactions

|http://www.ncbi.nlm.nih.gov/gap/PheGenIUCSC[Phenotype-Genotype Integrator]
|Merges GWAS Catalog Data with other NCBI databases to facilitate prioritisation of GWAS hits for follow-up

|http://targetvalidation.org[OpenTargets platform]
|Platform for the identification of drug targets using multiple data types including GWAS

|link:http://locuszoom.org/[LocusZoom]
|A suite of software for creating regional association plots from GWAS data

|https://bio.tools/?page=1&q=GWAS%20study&sort=score&ord=desc[Elixir Tools and Data Services Registry]
|A registry of GWAS-related resources and tools

|https://omictools.com/search?q=GWAS[OMICStools]
|Searchable database of GWAS and other bioinformatics tools


|===
// end::b-col-indv[]

=== Sources of summary statistics

Summary statistics are available via the https://www.ebi.ac.uk/gwas/downloads/summary-statistics[GWAS Catalog] where possible. Find additional sources of summary statistics https://www.ebi.ac.uk/gwas/downloads/summary-statistics[here].

=== Related projects

link:http://www.bristol.ac.uk/integrative-epidemiology/news/2016/data-mine.html[DATA MINE public art project]
