<!DOCTYPE html> <html lang="en"
                      xmlns:th="http://www.thymeleaf.org"
                      xmlns="http://www.w3.org/1999/xhtml">

<head>
    <meta charset="utf-8"/>
    <meta http-equiv="X-UA-Compatible" content="IE=edge"/>
    <meta name="viewport" content="width=device-width, initial-scale=1"/>
    <meta name="description"
          content="The NHGRI-EBI GWAS Catalog: a curated collection of all published genome-wide association studies, produced by a collaboration between EMBL-EBI and NHGRI"/>
    <meta name="keywords" content="GWAS Catalog, GWAS, NHGRI, EBI, EMBL-EBI, SPOT"/>
    <meta name="author" content="Tony Burdett, Emma Hastings, Dani Welter, SPOT, EMBL-EBI, NHGRI"/>
    <link rel="icon" href="../static/images/favicon.ico" th:href="@{images/favicon.ico}"/>

    <title>GWAS Catalog</title>

    <!-- Bootstrap core CSS -->
    <link href="../static/css/bootstrap.css" th:href="@{css/bootstrap.min.css}" rel="stylesheet"/>
    <link href="https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/themes/smoothness/jquery-ui.css"
          rel="stylesheet"/>
    <link href="../static/css/bootstrap-table.css" th:href="@{css/bootstrap-table.css}" rel="stylesheet"/>
    <link href="../static/css/bootstrap-theme.css" th:href="@{css/bootstrap-theme.min.css}" rel="stylesheet"/>

    <!-- Custom styles for this template -->
    <link href="../static/css/jumbotron.css" th:href="@{css/jumbotron.css}" rel="stylesheet"/>
    <link href="../static/css/sticky-footer-navbar.css" th:href="@{css/sticky-footer-navbar.css}" rel="stylesheet"/>

    <!-- Additional styling on top of bootstrap -->
    <link rel="stylesheet" href="../static/css/goci-ui.css" th:href="@{css/goci-ui.css}"/>
    <link rel="stylesheet" href="../static/css/ontology-autocomplete.css" th:href="@{css/ontology-autocomplete.css}"/>
    <link rel="stylesheet" href="../static/css/icons/flaticon.css" th:href="@{css/icons/flaticon.css}"/>
    <link rel="stylesheet" href="../static/css/goci-color-palette-1.css" th:href="@{css/goci-color-palette-1.css}"/>
    <link href="../static/css/gwas-icons" th:href="@{../css/gwas-icons.css}" rel="stylesheet"/>

    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
    <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->

    <script src="../static/js/console-plugin.js" th:src="@{js/console-plugin.js}"></script>

</head>

<body>


<div th:include="fragments/header :: navbar"></div>

<div th:include="fragments/header :: jumbotron"></div>

<div class="container-fluid">
    <ol class="breadcrumb background-color-complementary-accent">
        <li><a href="home">Home</a></li>
        <li class="active">10th Anniversary</li>
    </ol>
</div>

<div class="container tight-to-top">

    <div class="row">
        <h2 style="color: black">10th Anniversary of the NHGRI-EBI GWAS Catalog</h2>
        <br/>
        <p>The GWAS Catalog was launched in 2008 to provide a publicly available curated resource of all published human genome wide association studies (GWAS) and association results. Over 10 years, the Catalog has grown, from a single published GWAS on age-related macular degeneration (Klein et al., 2005), into the primary source of disease-related associations with genetic variants (watch the associations increase over time in this video). The Catalog now contains over 3,300 publications and almost 60,000 unique SNP-trait associations, with global reach and over 10,000 visitors per month. Extensive manual curation of GWA studies by expert scientists ensures that the Catalog provides an accurate, comprehensive and structured summary of GWAS results. This vast amount of data underpins research into common disease, enabling investigations to identify causal variants, understand disease mechanisms, and establish targets for novel therapies.</p>
        <br/>
        <p th:align="center"><video width="800" height="600" th:preload="preload" controls="controls">
            <source th:src="@{/video/gwas_catalog_2018_10th_anniversary.mp4}" type="video/mp4" />
            <source th:src="@{/video/gwas_catalog_2018_10th_anniversary.ogg}" type="video/ogg" />
            Your browser does not support the video tag.
        </video>
        </p><p>
        <br/>
        The GWAS landscape has evolved over the last 10 years, with developments in study design and genotyping technologies. GWAS have illustrated the highly polygenic nature of complex traits and the opportunity to address the issue of missing heritability (Manolio et al., 2009). The field has thus evolved with the examination of increasing sample sizes, number of genetic variants assayed and ever more specific traits, endophenotypes and measurements. Consequently, we have seen an increase in the complexity of GWAS: publications include multiple GWAS studies, combinatorial trait models, interaction effects, pleiotropy and multi-ancestry analysis. Here at the GWAS Catalog we adapt to ensure the resource contains the most relevant and up-to-date research that has the greatest utility to the scientific community.
        <br/>
        <br/>
        Highlights over the last 10 years include releasing a new GWAS Catalog website in 2015 (<a th:href="@{/home}">www.ebi.ac.uk/gwas</a>); mapping curated trait descriptions to ontology terms to enable enriched ontology-driven search capabilities; improving the well-known GWAS Catalog diagram; and more completely capturing ancestry. Accurate characterization of ancestry is essential to interpret human genomics data. To facilitate this we developed a framework to systematically describe detailed ancestry information (see our recent published framework and guidelines for reporting ancestry (Morales et al., 2018). Recently, with support from the scientific community, we started hosting summary statistics for GWAS Catalog studies. Integration of these full p-value datasets with the structured meta-data in the Catalog will facilitate community analysis and re-use of these valuable data. In the future, we will make these datasets more accessible via a comprehensive, searchable database, providing a platform for downstream meta-analysis. Please submit your data to us to extend the utility of your results and the potential for follow on studies.
        <br/>
        <br/>
        We are also investigating expanding our scope to include targeted arrays and going beyond array-based genotyping to include sequencing-based genotyping, which enables a deeper interrogation of rare disease-associated variants. With these developments come many challenges, including how best to represent the increasing complexity of study designs, trait architecture and statistical analyses, but with the help of GWAS Catalog publication authors and users we are excited about what the next 10 years of GWAS hold.
        <br/>
        <br/>
        <b>References</b>
        <br/>
        <br/>
        Klein RJ, Zeiss C, Chew EY, Tsai JY, Sackler RS, Haynes C, Henning AK, SanGiovanni JP, Mane SM, Mayne ST, Bracken MB, Ferris FL, Ott J, Barnstable C, Hoh J. Complement factor H polymorphism in age-related macular degeneration. Science. 2005 Apr 15;308(5720):385-9. Epub 2005 Mar 10. PubMed PMID: 15761122; PubMed Central PMCID: PMC1512523.
        <br/><br/>
        Manolio TA, Collins FS, Cox NJ, Goldstein DB, Hindorff LA, Hunter DJ, McCarthy MI, Ramos EM, Cardon LR, Chakravarti A, Cho JH, Guttmacher AE, Kong A, Kruglyak L, Mardis E, Rotimi CN, Slatkin M, Valle D, Whittemore AS, Boehnke M, Clark AG, Eichler EE, Gibson G, Haines JL, Mackay TF, McCarroll SA, Visscher PM. Finding the missing heritability of complex diseases. Nature. 2009 Oct 8;461(7265):747-53. doi: 10.1038/nature08494. Review. PubMed PMID: 19812666; PubMed Central PMCID: PMC2831613.
        <br/><br/>
        Morales J, Welter D, Bowler EH, Cerezo M, Harris LW, McMahon AC, Hall P, Junkins HA, Milano A, Hastings E, Malangone C, Buniello A, Burdett T, Flicek P, Parkinson H, Cunningham F, Hindorff LA, MacArthur JAL. A standardized framework for representation of ancestry data in genomics studies, with application to the NHGRI-EBI GWAS Catalog. Genome Biol. 2018 Feb 15;19(1):21. doi: 10.1186/s13059-018-1396-2. PubMed PMID: 29448949; PubMed Central PMCID: PMC5815218.
        <br/><br/>
    </p>


    </div>

</div>
<!-- /container -->

<div th:include="fragments/footer :: page_footer"></div>



<!-- Bootstrap core JavaScript
================================================== -->
<!-- Placed at the end of the document so the pages load faster -->
<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.2/jquery.min.js"></script>
<script src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/jquery-ui.min.js"></script>
<script src="../static/js/bootstrap.min.js" th:src="@{js/bootstrap.min.js}"></script>
<!-- IE10 viewport hack for Surface/desktop Windows 8 bug -->
<script src="../static/js/ie10-viewport-bug-workaround.js" th:src="@{js/ie10-viewport-bug-workaround.js}"></script>

<!-- GOCI JavaScript
================================================== -->
<script src="../static/js/goci-ui.js" th:src="@{../js/goci-ui.js}"></script>
<script src="../static/js/autocomplete-termselect.js" th:src="@{../js/autocomplete-termselect.js}"></script>
<script src="../static/js/jquery.autocomplete.js" th:src="@{../js/jquery.autocomplete.js}"></script>

<!-- Google analytics
================================================== -->
<script>
    (function(i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r;
        i[r] = i[r] || function() {
            (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date();
        a = s.createElement(o),
            m = s.getElementsByTagName(o)[0];
        a.async = 1;
        a.src = g;
        m.parentNode.insertBefore(a, m)
    })(window, document, 'script', '//www.google-analytics.com/analytics.js', 'ga');
    ga('create', 'UA-60195133-1', 'auto');
    ga('send', 'pageview');
</script>

</body>
</html>